prometheus/promauto/auto.go

```diff
@@ -79,51 +84,78 @@
-// metric, returning the registered metric, would result in nice code for those
-// using the method, but would pollute every single metric interface for
-// everybody avoiding the global registry.
-//
-// To address both issues, the problematic auto-registering and possibly
-// panicking constructors are all in this package with a clear warning
-// ahead. And whoever cares about avoiding global state and possibly panicking
-// function calls can simply ignore the existence of the promauto package
-// altogether.
-//
-// A final note: There is a similar case in the net/http package of the standard
-// library. It has DefaultServeMux as a global instance of ServeMux, and the
-// Handle function acts on it, panicking if a handler for the same pattern has
-// already been registered. However, one might argue that the whole HTTP routing
-// is usually set up closely together in the same package or file, while
-// Prometheus metrics tend to be spread widely over the codebase, increasing the
-// chance of surprising registration failures. Furthermore, the use of global
-// state in net/http has been criticized widely, and some avoid it altogether.
+// separate package?
+//
+// The main problem is that registration may fail, e.g. if a metric inconsistent
+// with the newly to be registered one is already registered. Therefore, the
+// Register method in the prometheus.Registerer interface returns an error, and
+// the same is the case for the top-level prometheus.Register function that
+// registers with the global registry. The prometheus package also provides
+// MustRegister versions for both. They panic if the registration fails, and
+// they clearly call this out by using the Must… idiom. Panicking is a bit
+// problematic here because it doesn't just happen on input provided by the
+// caller that is invalid on its own. Things are a bit more subtle here: Metric
+// creation and registration tend to be spread widely over the codebase. It can
+// easily happen that an incompatible metric is added to an unrelated part of
+// the code, and suddenly code that used to work perfectly fine starts to panic
+// (provided that the registration of the newly added metric happens before the
+// registration of the previously existing metric). This may come as an even
+// bigger surprise with the global registry, where simply importing another
+// package can trigger a panic (if the newly imported package registers metrics
+// in its init function). At least, in the prometheus package, creation of
+// metrics and other collectors is separate from registration. You first create
+// the metric, and then you decide explicitly if you want to register it with a
+// local or the global registry, and if you want to handle the error or risk a
+// panic. With the constructors in the promauto package, registration is
+// automatic, and if it fails, it will always panic. Furthermore, the
+// constructors will often be called in the var section of a file, which means
```

Comment:
That should be really not recommended. I would remove the word `often` .. (:

Reply:
A lot of people do use the global registry, and for them it's normal to create metrics in the `var` section. Even important players in the Prometheus community recommend that. We have to leave it open here what way to pick. If I remove the "often", it will sounds like you would _always_ create metrics in the `var` section.